src/query/subquery_cache.c

```diff
@@ -0,0 +1,513 @@
+
+  if (key == NULL)
+    {
+      return true;
+    }
+
+  sq_free_key (key);
+
+  ret = 0;
+  if (xasl->scan_ptr)
+    {
+      for (scan_ptr = xasl->scan_ptr; scan_ptr != NULL; scan_ptr = scan_ptr->next)
+	{
+	  ret |= sq_walk_xasl_check_not_caching (thread_p, scan_ptr);
+	}
+    }
+  if (xasl->aptr_list)
+    {
+      for (aptr = xasl->aptr_list; aptr != NULL; aptr = aptr->next)
+	{
+	  ret |= sq_walk_xasl_check_not_caching (thread_p, aptr);
+	}
+    }
+
+  return ret;
+}
+
+/*
+ * sq_cache_initialize () - Initializes the cache for a given XASL node.
+ *   return: NO_ERROR if successful, ER_FAILED otherwise.
+ *   xasl(in/out): The XASL node for which the cache is being initialized.
+ *
+ * This function creates a hash table for caching the results of the XASL node. It sets up initial values for cache hit and miss
+ * counters and marks the cache as initialized. The function returns NO_ERROR upon successful initialization, or ER_FAILED if the
+ * hash table could not be created.
+ */
+int
+sq_cache_initialize (THREAD_ENTRY * thread_p, XASL_NODE * xasl)
+{
+  UINT64 max_subquery_cache_size = (UINT64) prm_get_bigint_value (PRM_ID_MAX_SUBQUERY_CACHE_SIZE);
+  int sq_hm_entries = (int) max_subquery_cache_size / 2048;	// default 1024
+
+  xasl->sq_cache->ht = mht_create ("sq_cache", sq_hm_entries, sq_hash_func, sq_cmp_func);
+  if (!xasl->sq_cache->ht)
+    {
+      return ER_FAILED;
+    }
+
+  xasl->sq_cache->size_max = max_subquery_cache_size;
+  return NO_ERROR;
+}
+
+/*
+ * sq_put () - Puts a value into the cache for a given XASL node.
+ *   return: NO_ERROR if the value is successfully cached, ER_FAILED otherwise.
+ *   xasl(in): The XASL node for which the value is being cached.
+ *   regu_var(in): The regu variable containing the value to be cached.
+ *
+ * This function attempts to cache the result of a regu variable associated with a given XASL node. It generates a key based on
+ * the XASL node's structure and creates a cache entry if such a key does not already exist in the cache. The function returns
+ * NO_ERROR if the value is successfully cached, or ER_FAILED if the key could not be generated or the value could not be added
+ * to the cache.
+ */
+int
+sq_put (THREAD_ENTRY * thread_p, SQ_KEY * key, XASL_NODE * xasl, REGU_VARIABLE * regu_var)
+{
+  SQ_VAL *val;
+  const void *ret;
+  UINT64 new_entry_size = 0;
+  int i;
+
+  val = sq_make_val (thread_p, regu_var);
+
+  if (!xasl->sq_cache->ht)
+    {
+      sq_cache_initialize (thread_p, xasl);
+    }
+
+  for (i = 0; i < key->n_elements; i++)
+    {
+      new_entry_size += (UINT64) or_db_value_size (key->dbv_array[i]);
+    }
+  new_entry_size += sizeof (SQ_KEY);
+  switch (val->t)
+    {
+    case TYPE_CONSTANT:
+      new_entry_size += (UINT64) or_db_value_size (val->val.dbvalptr) + sizeof (SQ_VAL);
+      break;
+
+    default:
+      break;
+    }
+
+  if (xasl->sq_cache->size_max < xasl->sq_cache->size + new_entry_size)
+    {
+      XASL_CLEAR_FLAG (xasl, XASL_SQ_CACHE);
```

Comment:
If the information set when generating xasl is changed during execution, it must be restored at the end.

Here, the `XASL_SQ_CACHE` flag must not be cleared.
If the flag is cleared to disable the subquery cache, the flag must be set again at the end.

Please refer to the reproducible scenario.
Do not use recompile hint.
```
drop table if exists t1, t2;
create table t1 (c1 int, c2 int);
create table t2 (c1 int, c2 int);

insert into t1 values (1, 1), (2, 2);
insert into t2 values (1, 10), (2, 20);

set trace on;

select t1.c2, (select t2.c2 from t2 where t2.c1 = t1.c1) from t1;
show trace;
/* SUBQUERY_CACHE (hit: 0, miss: 2, size: 96 Bytes, status: enabled) */

select t1.c2, (select t2.c2 from t2 where t2.c1 = t1.c1) from t1;
show trace;
/* nothing */
```

Reply:
I've handled that error. Please confirm.

Reply:
(I added `bool enabled` in sq_cache struct)